volumes/miovision/sql/function/function-aggregate_volumes_daily.sql

```diff
@@ -0,0 +1,126 @@
+	start_date date,
+	end_date date
+)
+RETURNS void
+LANGUAGE 'plpgsql'
+
+COST 100
+VOLATILE
+AS $BODY$
+
+DECLARE n_deleted int;
+DECLARE n_inserted int;
+
+BEGIN
+    
+    WITH deleted AS (
+        DELETE FROM miovision_api.volumes_daily
+        WHERE 
+            dt >= start_date
+            AND dt < end_date
+        RETURNING *
+    )
+    
+    SELECT COUNT(*) INTO n_deleted
+    FROM deleted;
+
+    RAISE NOTICE 'Deleted % rows from miovision_api.volumes_daily for % to %.', n_deleted, start_date, end_date;
+ 
+    --identify duration of unacceptable gaps per day, intersection
+    WITH unacceptable_gaps_summarized AS (
+        SELECT
+            un.intersection_uid,
+            avg_vols.classification_uid,
+            un.datetime_bin::date AS gap_date,
+            SUM(un.gap_minutes_15min) AS unacceptable_gap_minutes,
+            SUM(un.gap_minutes_15min * avg_vols.avg_hour_vol)/60 AS avg_historical_gap_vol
+        FROM miovision_api.unacceptable_gaps AS un
+        --find average historical volume by classification during those gaps
+        LEFT JOIN miovision_api.gapsize_lookup AS avg_vols ON
+            avg_vols.intersection_uid = un.intersection_uid
+            AND avg_vols.hour_bin = date_part('hour', un.datetime_bin)
+            AND avg_vols.dt = un.dt
+            AND avg_vols.classification_uid IS NOT NULL
+        WHERE
+            un.datetime_bin::date >= start_date
+            AND un.datetime_bin::date < end_date
+        GROUP BY
+            un.intersection_uid,
+            un.datetime_bin::date,
+            avg_vols.classification_uid
+    ),
+    
+    inserted AS (
+        INSERT INTO miovision_api.volumes_daily (
+            intersection_uid, dt, classification_uid, daily_volume, isodow,
+            holiday, unacceptable_gap_minutes, datetime_bins_missing, avg_historical_gap_vol
+        )
+        SELECT
+            v.intersection_uid,
+            v.datetime_bin::date AS dt,
+            v.classification_uid,
+            SUM(v.volume) AS daily_volume,
+            date_part('isodow', v.datetime_bin::date) AS isodow,
+            hol.holiday IS NOT NULL AS holiday,
+            COALESCE(un.unacceptable_gap_minutes, 0) AS unacceptable_gap_minutes,
+            60 * 24 - COUNT(DISTINCT datetime_bin) AS datetime_bins_missing,
+            un.avg_historical_gap_vol::int
+        --raw data
+        FROM miovision_api.volumes AS v
+        LEFT JOIN ref.holiday AS hol ON hol.dt = v.datetime_bin::date
+        --identify duration of unacceptable gaps (extended zero periods)
+        --and avg historical volumes by classification for those gaps
+        LEFT JOIN unacceptable_gaps_summarized AS un ON
+            un.intersection_uid = v.intersection_uid
+            AND v.datetime_bin::date = un.gap_date
+            AND un.classification_uid = v.classification_uid
+        --anti join anomalous_ranges table
+        LEFT JOIN miovision_api.anomalous_ranges AS ar ON
```

Comment:
so technically everytime we update anomalous range we will have to update this daily table? 🤔

Reply:
The only solution I can think of to that is have a view on top of this table that filters them out. 
Or... a function/trigger to use when inserting to anomalous ranges? :/

Reply:
I would like it to be something that a user can't easily ignore, and not put it on them to be joining

Reply:
having a view on top of this table like gabe suggested sounds easier and less error prone. I think we should probably start announcing newly identified anomalous range and the potential changed on the view?

Like a dag task that looks at newly added anomalous ranges and :Cat_yell: in an alert message tagging daily table users?

Reply:
Two workable options: 
1. new view which takes into account `anomalous_ranges` table
2. I could add an optional intersection parameter to the daily agg function and then using an on insert / update / delete trigger on the `anomalous_ranges` table we could re-run the daily aggregation for any affected intersections. 
3. ~~Let the user exclude a_r~~

Reply:
How do you feel about table `volumes_daily_donotuse` and view `volumes_daily`, plus some comments?

Reply:
How about `volumes_daily` vs `volumes_daily_unfiltered`? I think there could be valid reasons for wanting to bypass the `anomalous_ranges` filter, such as for QA.

Reply:
_unfiltered sounds good. As long as we restrict bdit_humans select access to it + give investigators select access I think it should be good :meow_salute: